---
title: "Patient 1315 scRNA analyisis"
output: github_document
---

##Methods and Data Processing:
```{r init, echo = FALSE}
library(dplyr)
library(patchwork)
library(Seurat)
library(tidyverse)
library(Matrix)
library(scales)
library(SingleR)
library(pheatmap)
```

#load data
```{R}
# Create each individual Seurat object for every sample
for (file in c("PT1315_Normal_ft_bc_mx","PT1315_Tumor_ft_bc_mx")){
  seurat_data_1315 <- Read10X(data.dir = paste0("pt1315/", file))
  seurat_obj_pt1315 <- CreateSeuratObject(counts = seurat_data_1315,
                                   min.cells = 20,
                                   min.features = 200, 
                                   project = file)
  assign(file, seurat_obj_pt1315)
}
seurat_obj_pt1315 
```
```{r}
#check the metadata in the new seurat objects
PT1315_Normal_ft_bc_mx
```
```{r}
PT1315_Tumor_ft_bc_mx
```
#create a merged seurat object
```{r}
merged_seurat_pt1315 <- merge(x = PT1315_Normal_ft_bc_mx, 
                       y = PT1315_Tumor_ft_bc_mx, 
                       add.cell.id = c("Normal", "Tumor"),
                       project = "Serious_Tumor_pt1272") #Endometrial carcinosarcoma

merged_seurat_pt1315 #check the merged object
```

```{r}
#examining first 30 cells, genes known to be expressed in a good number of cells 
seurat_data_1315[c("CD3D","TCL1A","MS4A1"), 1:30]
```

```{r}
dense.size_1315 <- object.size(as.matrix(seurat_data_1315))
dense.size_1315
sparse.size_1315 <- object.size(seurat_data_1315)
sparse.size_1315
dense.size_1315 / sparse.size_1315
```
```{r }
###add number if genes per UMI for each cell to metadata
merged_seurat_pt1315$log10GenesPerUMI <- log10(merged_seurat_pt1315$nFeature_RNA) / log10(merged_seurat_pt1315$nCount_RNA)

```
#compute percent mito ratio 
```{r}
#percentfeatureset will take a pattern and search the gene indentifiers,
#for each column (cell) it will take the sum of the counts slot for features belonging to the set, 
#divide by the column sum for all features and multiply by 100. ("-MT") works for human genes
merged_seurat_pt1315$mitoRatio <- PercentageFeatureSet(object = merged_seurat_pt1315, pattern = "^MT-")
merged_seurat_pt1315$mitoRatio <- merged_seurat_pt1315@meta.data$mitoRatio / 100 #want the ratio value for plotting, reversing step and then divinding by 100

```
###plot merged_seurat histogram 
```{r}
#most cells have less than 0.1 percent mitochondrial RNA
hist(merged_seurat_pt1315[["mitoRatio"]]$mitoRatio , breaks = 20) ######
```

```{r}
head(merged_seurat_pt1315)
```
#Visualize QC metrics as a violin plot
```{r}
VlnPlot(merged_seurat_pt1315, features = c("nFeature_RNA", "nCount_RNA", "mitoRatio"), ncol = 3)
# FeatureScatter is typically used to visualize feature-feature relationships, but can be used for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

FeatureScatter(merged_seurat_pt1315, feature1 = "nCount_RNA", feature2 = "mitoRatio")
FeatureScatter(merged_seurat_pt1315, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")

plot1_1315 + plot2_1315

plot.ecdf( merged_seurat_pt1315[["nCount_RNA"]]$nCount_RNA )
merged_seurat_pt1315 <- subset(merged_seurat_pt1315, subset = nFeature_RNA < 2000 & nCount_RNA < 5000 & mitoRatio < 5)

```
```{r}
#create metadata dataframe 
metadata_pt1315 <- merged_seurat_pt1315@meta.data

# add cell IDs to metadata
metadata_pt1315$cells <-rownames(metadata_pt1315)

#rename columns 
metadata_pt1315 <- metadata_pt1315 %>%
  dplyr::rename(seq_folder = orig.ident,
                nUMI = nCount_RNA,
                nGene = nFeature_RNA)
```
###create metadata fr
```{r}
#create sample column

metadata_pt1315$sample <- NA
metadata_pt1315$sample[which(str_detect(metadata_pt1315$cells, "^Normal_"))] <- "Normal"
metadata_pt1315$sample[which(str_detect(metadata_pt1315$cells, "^Tumor_"))] <- "Tumor"

#add metadata back to seurat object
merged_seurat_pt1315@meta.data <- metadata_pt1315

#create .RDaTa object to load at any time
saveRDS(merged_seurat_pt1315, file = "pt1315/merged_filtered_pt1315.RData")
```

```{r}
#visualize the number of cell counts per sample
metadata_pt1315 %>% 
  ggplot(aes(x= sample, fill= sample)) +
  geom_bar() +
  theme_classic()+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust= 0.05, face= "bold")) +
  ggtitle("NCells")

```

```{r}
#visualize the number of UMIs/transcripts per cell
metadata_pt1315 %>%
  ggplot(aes(color = sample, x= nUMI, fill= sample)) +
  geom_density(alpha = 0.2) + 
  scale_x_log10(labels = comma) + 
  theme_classic() +
  ylab("Cell density") +
  geom_vline(xintercept = 500)
```

```{r}
#visualize the distribution of genes detected per cell via histogram
metadata_pt1315 %>% 
  ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  geom_density(alpha = 0.2) + 
  theme_classic() +
  scale_x_log10() + 
  geom_vline(xintercept = 300)

```

```{r}
#visualize the distribution of genes detected per cell via boxplot
metadata_pt1315 %>% 
  ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  geom_boxplot() + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold")) +
  ggtitle("NCells vs NGenes")

```

```{r}
#visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
metadata_pt1315 %>% 
  ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  geom_point() + 
  scale_colour_gradient(low = "gray90", high = "black") +
  stat_smooth(method=lm) +
  scale_x_log10() + 
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = 500) +
  geom_hline(yintercept = 250) +
  facet_wrap(~sample)
```
```{r}
#visualize the distribution of mitochondrial gene expression detected per cell
metadata_pt1315 %>% 
  ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  geom_density(alpha = 0.2) + 
  scale_x_log10() + 
  theme_classic() +
  geom_vline(xintercept = 0.2)
```
```{r}
# Visualize the overall complexity of the gene expression by visualizing the genes detected per UMI
metadata_pt1315 %>%
  ggplot(aes(x=log10GenesPerUMI, color = sample, fill=sample)) +
  geom_density(alpha = 0.2) +
  theme_classic() +
  geom_vline(xintercept = 0.8)
```
#cell-level filtering
#filter out low quality reads 
```{r}
filtered_seurat_1315 <- subset(x= merged_seurat_pt1315,
                          subset= (nUMI >= 500) &
                            (nGene >=250) &
                            (log10GenesPerUMI > 0.80) &
                            (mitoRatio < 0.20))
```



#gene-level filtering
remove genes that have zero expression in all cells and perform some filtering by prevalance. 

```{r}
#extract counts

countspt1315 <- GetAssayData(object = filtered_seurat_1315, slot = "counts")
#output a logical vector for every gene on whether it has more than zero counts per cell
nonzeropt1315 <- countspt1315 > 0

#sums all TRUE values and returns TRUE if more than 10 TRUE values per gene
keep_genespt1315 <- Matrix::rowSums(nonzeropt1315) >= 10

#keep only genes expressed in more than 10 cells
filtered_counts_pt1315 <- countspt1315[keep_genespt1315, ]

#update the filtered counts in the filtered_seurat object
filtered_seurat_1315[["RNA"]] <- CreateAssayObject(counts = filtered_counts_pt1315)

#save filtered subset to new metadata (optional, depending on your use case)
metadata_clean_pt1315 <- filtered_seurat_1315@meta.data

```
##re-assesing QC metrics 
performing same plots as with the unfiltered sata to determine if threshhods where appropiate

```{r}
#visualize the number of cell counts per sample
metadata_clean_pt1315 %>% 
  ggplot(aes(x= sample, fill= sample)) +
  geom_bar() +
  theme_classic()+
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust= 0.05, face= "bold")) +
  ggtitle("NCells")

```

```{r}
#visualize the nu,ber of UMIs/transcripts per cell
metadata_clean_pt1315 %>%
  ggplot(aes(color = sample, x= nUMI, fill= sample)) +
  geom_density(alpha = 0.2) + 
  scale_x_log10() + 
  theme_classic() +
  ylab("Cell density") +
  geom_vline(xintercept = 500)
```

```{r}
#visualize the distribution of genes detected per cell via histogram
metadata_clean_pt1315 %>% 
  ggplot(aes(color=sample, x=nGene, fill= sample)) + 
  geom_density(alpha = 0.2) + 
  theme_classic() +
  scale_x_log10() + 
  geom_vline(xintercept = 300)
```

```{r}
#visualize the distribution of genes detected per cell via boxplot
metadata_clean_pt1315 %>% 
  ggplot(aes(x=sample, y=log10(nGene), fill=sample)) + 
  geom_boxplot() + 
  theme_classic() +
  theme(axis.text.x = element_text(angle = 45, vjust = 1, hjust=1)) +
  theme(plot.title = element_text(hjust=0.5, face="bold")) +
  ggtitle("NCells vs NGenes")
```

```{r}
#visualize the correlation between genes detected and number of UMIs and determine whether strong presence of cells with low numbers of genes/UMIs
metadata_clean_pt1315 %>% 
  ggplot(aes(x=nUMI, y=nGene, color=mitoRatio)) + 
  geom_point() + 
  scale_colour_gradient(low = "gray90", high = "black") +
  stat_smooth(method=lm) +
  scale_x_log10() + 
  scale_y_log10() + 
  theme_classic() +
  geom_vline(xintercept = 500) +
  geom_hline(yintercept = 250) +
  facet_wrap(~sample)
```

```{r}

#visualize the distribution of mitochondrial gene expression detected per cell
metadata_clean_pt1315 %>% 
  ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  geom_density(alpha = 0.2) + 
  scale_x_log10() + 
  theme_classic() +
  geom_vline(xintercept = 0.2)
```

```{r}
#visualize the distribution of mitochondrial gene expression detected per cell
metadata_clean_pt1315 %>% 
  ggplot(aes(color=sample, x=mitoRatio, fill=sample)) + 
  geom_density(alpha = 0.2) + 
  scale_x_log10() + 
  theme_classic() +
  geom_vline(xintercept = 0.2)

```
##count Normalization and PCA

```{r}
#normalize the counts
seurat_phase_pt1315 <- NormalizeData(filtered_seurat_1315)

```
# Normalize the counts and perform cell cycle scoring

```{r}
#get the cell cycle marker genes list from Seurat
data("cc.genes")
#extract the G2/M phase and S phase marker genes
g2m_genes_1315 <- cc.genes$g2m.genes
s_genes_1315 <- cc.genes$s.genes
```

```{r}
#score cells for cell cycle
#recommended to cceck the cell cycle phase before perfoming sctransfrom method
seurat_phase_pt1315 <- CellCycleScoring(seurat_phase_pt1315, 
                                 g2m.features = g2m_genes_1315, 
                                 s.features = s_genes_1315)
```

```{r}
#view cell cycle scores and phases assigned to cells                                 
View(seurat_phase_pt1315@meta.data)
```

```{r}
#identify the most variable genes
seurat_phase_pt1315 <- FindVariableFeatures(seurat_phase_pt1315, 
                                     selection.method = "vst",
                                     nfeatures = 2000, 
                                     verbose = FALSE)
```
# Scale the counts
```{r}
seurat_phase_pt1315 <- ScaleData(seurat_phase_pt1315)
```
#PCA
```{r}
seurat_phase_pt1315 <- RunPCA(seurat_phase_pt1315)
```
# Plot the PCA colored by cell cycle phase
```{r}
DimPlot(seurat_phase_pt1315,
        reduction = "pca",
        group.by= "Phase",
        split.by = "Phase")
```
#SCTransform
-more accurate method of normalizing, estimating the variance of the raw filtered data, and indentifying the most variable genes
-by default, sctranform accounts for cellular sequencing depth, or nUMIs
- SCTransform can be applied directly on the merged object.
```{r}
#split seurat object by condition to perform cell cycle scoring and SCT on all samples
split_seurat_pt1315 <- SplitObject(filtered_seurat_1315, split.by = "sample")

split_seurat_pt1315 <- split_seurat_pt1315[c("Normal", "Tumor")]

for (i in 1:length(split_seurat_pt1315)) {
  split_seurat_pt1315[[i]] <- NormalizeData(split_seurat_pt1315[[i]], verbose = TRUE)
  split_seurat_pt1315[[i]] <- CellCycleScoring(split_seurat_pt1315[[i]], g2m.features=g2m_genes_1315, s.features=s_genes_1315)
  split_seurat_pt1315[[i]] <- SCTransform(split_seurat_pt1315[[i]], vars.to.regress = c("mitoRatio"))
}

#check which assays are stored in objects
split_seurat_pt1315$Normal@assays
split_seurat_pt1315$Tumor@assays


```

##integration
```{r}
#select the most variable features to use for integration
integ_features_1315 <- SelectIntegrationFeatures(object.list = split_seurat_pt1315, 
                                            nfeatures = 3000) 
#prepare the SCT list object for integration
split_seurat_pt1315 <- PrepSCTIntegration(object.list = split_seurat_pt1315, 
                                   anchor.features = integ_features_1315)
#integration anchors, find best buddies
integ_anchors_pt1315 <- FindIntegrationAnchors(object.list = split_seurat_pt1315, 
                                        normalization.method = "SCT", 
                                        anchor.features = integ_features_1315)
#integrate across conditions
seurat_integrated_pt1315 <- IntegrateData(anchorset = integ_anchors_pt1315, 
                                   normalization.method = "SCT")
```
#UMAP visualization
```{r}
#run PCA
seurat_integrated_pt1315 <- RunPCA(object = seurat_integrated_pt1315)

```

```{r}
#plot PCA
PCAPlot(seurat_integrated_pt1315,
        split.by = "sample",
        group.by = "sample")  
```

```{r}
# Run UMAP
seurat_integrated_pt1315 <- RunUMAP(seurat_integrated_pt1315, 
                             dims = 1:40,
                             reduction = "pca")
```

```{r}
# Plot UMAP                             
DimPlot(seurat_integrated_pt1315, group.by = "sample") 

```

```{r}
#sometime easier to see whther all of the cells align well if we split the plotting between conditions
DimPlot(seurat_integrated_pt1315,
        split.by = "sample",
        group.by = "sample")  
```
##clustering cells based on top PCs

```{r}
#printing out the most variable genes driving PCs
print(x = seurat_integrated_pt1315[["pca"]], 
      dims = 1:10, 
      nfeatures = 5)
```

```{r}
# Plot the elbow plot
ElbowPlot(object = seurat_integrated_pt1315,ndims = 30)
```

##Cluster cells 

```{r}
#determine the K-nearest neighbor graph
seurat_integrated_pt1315 <- FindNeighbors(object = seurat_integrated_pt1315, 
                                   dims = 1:30)

#determine the clusters for various resolutions                                
seurat_integrated_pt1315 <- FindClusters(object = seurat_integrated_pt1315,
                                  resolution = c(0.1, 0.3, 0.5, 1.0, 1.4))
#explore resolutions
seurat_integrated_pt1315@meta.data %>% 
  View()

DimPlot(seurat_integrated_pt1315, group.by = "integrated_snn_res.0.3", label = TRUE, label.size = 6)

```

# Assign identity of clusters
```{r}
Idents(seurat_integrated_pt1315) <- "integrated_snn_res.0.3"
```
##visulaze cell clusters
```{r}

DimPlot(seurat_integrated_pt1315,
        reduction = "umap",
        label = TRUE,
        label.size = 6)

```
###run umap
```{r}
# Run UMAP
seurat_integrated_pt1315 <- RunUMAP(seurat_integrated_pt1315, 
                             dims = 1:30,
                             )
```

##visulaze cell clusters

```{r}

# Plot the UMAP
DimPlot(seurat_integrated_pt1315,
        reduction = "umap"
        )

```

###exploration of quality control metrics
-can be useful ti explore these metrics visually to see if any clusters exhibit
-enrichment or are from different from other clusters
```{r}

n_cells_1315 <- FetchData(seurat_integrated_pt1315, 
                     vars = c("ident", "seq_folder")) %>%
  dplyr::count(ident, seq_folder) %>%
  tidyr::spread(ident, n)

# View table
View(n_cells_1315)
```
#visualize the cells per cluster for each sample usiNg UMAP

```{r}
#UMAP of cells in each cluster by sample
DimPlot(seurat_integrated_pt1315, 
        label = TRUE, 
        split.by = "sample")  + NoLegend()
```
#segregation of clusters by cell cycle phase

```{r}
#explore whether clusters segregate by cell cycle phase
DimPlot(seurat_integrated_pt1315,
        label = TRUE, 
        split.by = "Phase")  + NoLegend()
```
#explore additional metrics: number of UMIs and genes per cell, S-phase and G2M-phase markers, and mitochondrial gene expression by UMAP
```{r}
#determine metrics to plot present in seurat_integrated@meta.data
metrics_1315 <-  c("nUMI", "nGene", "S.Score", "G2M.Score", "mitoRatio")

FeaturePlot(seurat_integrated_pt1315, 
            reduction = "umap", 
            features = metrics_1315,
            pt.size = 0.4, 
            sort.cell = TRUE,
            min.cutoff = 'q10',
            label = TRUE)

```
#Exploration of the PCs driving the different clusters
```{r}
# Defining the information in the seurat object of interest
columns_1315 <- c(paste0("PC_", 1:16),
            "ident",
            "UMAP_1", "UMAP_2")

# Extracting this data from the seurat object
pc_data_1315 <- FetchData(seurat_integrated_pt1315, 
                     vars = columns_1315)

umap_label_1315 <- FetchData(seurat_integrated_pt1315, 
                        vars = c("ident", "UMAP_1", "UMAP_2"))  %>%
  group_by(ident) %>%
  summarise(x=mean(UMAP_1), y=mean(UMAP_2))
  
#plotting a UMAP plot for each of the PCs
map(paste0("PC_", 1:16), function(pc){
        ggplot(pc_data_1315, 
               aes(UMAP_1, UMAP_2)) +
                geom_point(aes_string(color=pc), 
                           alpha = 0.7) +
                scale_color_gradient(guide = FALSE, 
                                     low = "grey90", 
                                     high = "blue")  +
                geom_text(data=umap_label_1315, 
                          aes(label=ident, x, y)) +
                ggtitle(pc)
}) %>% 
        plot_grid(plotlist = .)

```
# Examine PCA results

```{r}
print(seurat_integrated_pt1315[["pca"]], dims = 1:5, nfeatures = 5)
```
###Exploring known cell type markers
```{r}
DimPlot(object = seurat_integrated_pt1315, 
        reduction = "umap", 
        label = TRUE) + NoLegend()

```

##cell annotation 

```{r}
data_1315 <- seurat_integrated_pt1315
DimPlot(data_1315, reduction = "umap", label= TRUE)

ref_1315 <- celldex::HumanPrimaryCellAtlasData()

results_1315 <- SingleR(test= as.SingleCellExperiment(data_1315), ref =ref_1315, labels = ref_1315$label.main)
results_1315

data_1315$singler_labels <- results_1315$labels

data_1315[[]]

DimPlot(data_1315, reduction = 'umap', group.by = 'singler_labels', label = TRUE)

```

###annotation diagnostics

```{r}
#based on the scores within cells
results_1315
results_1315$scores
```

```{r}
###plotscoreheat map

plotScoreHeatmap(results_1315) 
```
### based on deltas across cells
plot the ditrubution of delta values. each cell gets assigned a delta value, low delta vlaues indicate that the assginemt is uncertain. the idea is to detect the cells that are outliers within each labels which could indicate ambiguous assigminmnets. 
```{r}
plotDeltaDistribution(results_1315)
```
### comparing to unsupervised clustering 

#compare the distribution of cells across different clusters. calculate the numbers of cell type labels of each cluster. Perform clustering

```{r}

tab_1315 <- table(Assigned = results_1315$labels, Clusters = data_1315$integrated_snn_res.0.3)
tab_1315

```

```{r}
###create a heat map
#log tranform table values. add pseudo count of 10. 


pheatmap(log10(tab_1315 + 10), color = colorRampPalette(c('white', 'blue'))(10))

# Save final R object
saveRDS(seurat_integrated_pt1315, file = "pt1315/seurat_labeled_1315.rds")

```

# Save final R object


```{r}
saveRDS(seurat_integrated_pt1315, file = "pt1315/seurat_labeled_1315.rds")
```

**Session Info**
  
```{r}
sessionInfo()
```
</details>   
